connector/connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamHandler.scala

```diff
@@ -114,10 +123,97 @@ class SparkConnectStreamHandler(responseObserver: StreamObserver[Response]) exte
       responseObserver.onNext(response.build())
     }
 
-    responseObserver.onNext(sendMetricsToResponse(clientId, rows))
+    responseObserver.onNext(sendMetricsToResponse(clientId, dataframe))
     responseObserver.onCompleted()
   }
 
+  def processRowsAsArrowBatches(clientId: String, dataframe: DataFrame): Unit = {
+    val spark = dataframe.sparkSession
+    val schema = dataframe.schema
+    // TODO: control the batch size instead of max records
```

Comment:
I think you can remove this since we're already handling the max records?

Reply:
https://github.com/apache/spark/pull/38468#discussion_r1013186031 suggested control the batch size < 4MB

Reply:
interesting. I only know that gRPC has a hard limits of 2GB/s transfer rate. Never know it might not favor over large messages.

Reply:
I don't think there is a throughput limit in GRPC itself. 

The reason for the batching is that protobuf is not suited for this. Embedding large binary objects might require the reader to materialize them in memory. 

Fixing this is an optimization for later.

Reply:
Another downside of large allocations is that the GC does not really like them. All large allocation (> 1 MB) are generally placed in the old generation immediately, which requires a full GC to clean-up.